README.md
# Server to Server GTM Client

![logo](https://github.com/square/server-to-server-gtm-client/blob/main/img/logo-96.png)

A Server-Side Google Tag Manager client template which accepts HTTP POST requests and parses the request body into events. Useful when sending data to sGTM from another server such as your customer data platform.

## Recommendations

- When sending data to sGTM using this client, format your event structure to match Google's [Common Event Data](https://developers.google.com/tag-manager/serverside/common-event-data). This event API is used by most tags.

- When configuring tags in sGTM, Always include tag name under **Advanced Settings** in the tag configuration. This will aid in the tag error logging that this client supports.

![tag metadata form](https://github.com/square/server-to-server-gtm-client/blob/main/img/metadata.png)

- If also using a standard client from Google Marketing Platform, we recommend setting the Server-to-Server as a higher priority and requiring an authorization header. This will ensure requests intended for the Server-To-Server client are appropriately claimed and requests without an authorization header are then passed on to the next client to be evaluated. Be sure to check **Reject unauthorized requests** to ensure any requests that include an incorrect authorization header are still claimed and rejected by the Server-to-Server client.

![client priority](https://github.com/square/server-to-server-gtm-client/blob/main/img/priority.png)

![recommended auth settings](https://github.com/square/server-to-server-gtm-client/blob/main/img/require-token.png)



## Request Body Structure

Requests accepted by this client must have a JSON body that parses into an array of objects. Each object represents an event. Even if you are only sending a single event object, it must be contained in an array. Note that if you are sending multiple events, ensure your request payload size is smaller than 1mb. Any larger and Server-Side Google Tag Manager will return a 413 error.

Here is an example of a valid minimum request body.
```json
[
  {
    "event_name": "purchase"
  }
]
```

Another example with more fields and multiple events.
```json
[
  {
    "event_name": "purchase",
    "ip_override": "1.2.3.4",
    "language": "en_us"
  },
  {
    "event_name": "add_to_cart",
    "currency": "USD",
    "value": 7.77,
    "language": "en_es"
  },
  {
    "event_name": "login",
    "user_id": "FI4L21234",
    "viewport_size": "725x345",
    "user_data": {
      "email_address": "foo@example.com",
      "phone_number": "+15551234567",
      "first_name": "Jane",
      "last_name": "Doe"
    }
  }
]
```

Once the client is set up, you can test sending an event with cURL

```bash
curl -X POST https://sgtm.appspot.com \
  -H 'Content-Type: application/json' \
  -d '[{ "event_name": "purchase" }]'
```


## Authorization

This client template provides the option of requiring an authorization header for requests. This is an option for regular requests as well as preview requests. Add a bearer token in the client configuration UI. 

![client config](https://github.com/square/server-to-server-gtm-client/blob/main/img/authorization.png)

Then, include an authorization header in your requests:

### Production

```bash
curl -X POST https://sgtm.appspot.com \
  -H 'Authorization: Bearer prod-token' \
  -H 'Content-Type: application/json' \
  -d '[{ "event_name": "purchase" }]'
```


### Preview

If requiring an authorization header in preview mode, Be sure to also include the `X-Gtm-Server-Preview HTTP header` obtained from the preview UI by clicking **Send Requests Manually**.

![preview ui](https://github.com/square/server-to-server-gtm-client/blob/main/img/preview-ui.png)


```bash
curl -X POST https://sgtm.appspot.com \
  -H 'Authorization: Bearer preview-token' \
  -H 'x-gtm-server-preview: PASTE_PREVIEW_HEADER_HERE' \
  -H 'Content-Type: application/json' \
  -d '[{ "event_name": "purchase" }]'
```

## Type Checking

By default, this client type checks all [Common Event Data](https://developers.google.com/tag-manager/serverside/common-event-data) fields in each event. `event_name` is a required field, but all others are optional. Additional fields can be added or existing fields' types or required status can be overwritten in the **Type Checking** section of the client configuration UI.

## Primitive Types

If only adding checking for types `string`, `number`, or `boolean`, use the primitive type checking table. 

![primitive types table](https://github.com/square/server-to-server-gtm-client/blob/main/img/primitive.png)

## Structural Types

If type checking arrays or objects, use the structural types input. Enter a JSON string which will be added to the schema. The client configuration UI will not indicate if your JSON is valid or not; ensure this before entering a value.

See the examples below for how to add type checking.

![structural types table](https://github.com/square/server-to-server-gtm-client/blob/main/img/structural.png)


### Objects

The `value` field below contains the schema of the `locale` object. Any fields in the locale object can also be of type `array` or `object` and contain further nested schema.

```json
{
  "locale": {
    "type": "object",
    "required": false,
    "value": {
      "countryCode": { "type": "string", "required": false },
      "language": { "type": "string", "required": false },
    }
  }
}
```

### Array of primitives

The `value` field below contains the schema of the `categories` array. In this case, `categories` is an array of strings. Note that all elements in an array must be of the same type.

```json
{
  "categories": {
    "type": "array",
    "required": false,
    "value": {
      "type": "string",
      "required": false
    }
  }
}
```

### Array of objects

The `value` field below contains the schema of the `people` object. The `people` object has it's own schema specified in its `value` field.  Note that all objects in an array must be of the same schema.

```json
{
  "people": {
    "type": "array",
    "required": false,
    "value": {
      "type": "object",
      "required": false,
      "value": {
        "name": { "type": "string", "required": false },
        "height": { "type": "string", "required": false },
        "age": { "type": "number", "required": false }
      }
    }
  }
}
```

### Primitives

You can also use this input box to enter type checking for primitives, though it is easier to use the table above. Primitives do not require a `value` field, as there is no further nested type checking.

```json
{
  "timestamp": { "type": "string", "required": "true" }
}
```


## Client Response

This client will claim a request if it includes the correct authorization (or no authorization if none is required). If a request is accepted, the client will return one of the following responses.

### Response Codes

| Response Code | Error | Client Action |
|-|-|-|
| 200 | No error | claim request, respond success |
| 206 | A few tags or events fail but others succeed | claim request, respond partial success | 
| 400 | Invalid body, not json | claim request, respond error |
| 401 | Invalid authorization header | claim request, respond error |
| 422 | Type errors | claim request, respond error |
| 500 | Events fit the schema but all tags fail | claim request, respond error |


### Response Body

In addition, the client will include the following JSON response body

|Key|Type|Description|
|-|-|-|
|`status`|Integer|The status code|
|`error`|String|Error message. If the status of the response is 200, this field will be an empty string.|
|`invalidSchemaEvents`|Array of Integers|The list of events with an invalid schema. The integers point to the index of the event which failed. The order of the events is the same in the request and the response.|
|`fullyFailedTagEvents`|Array of Integers|The list of events where all tags that were triggered by the event failed. The integers point to the index of the event which failed. The order of the events is the same in the request and the response.|
|`partialFailedTagEvents`|Array of Integers|The list of events where some but not all tags that were triggered by the event failed. The integers point to the index of the event which failed. The order of the events is the same in the request and the response.|
|`failedTags`|Array of Objects|A list of objects where each object has two properties. `tagInfo`, the information about the tag that failed and `eventIndex`, the index of the event which failed.|


#### Example JSON

```json
{
  "status": 206,
  "error": "[GTM-XXXXXX][Server-to-Server Client][Error] - Partial Success: At least one event had the correct schema and at least one tag executed successfully. However, some failed.",
  "invalidSchemaEvents": [ 1, 3 ],
  "fullyFailedTagEvents": [ 0, 2 ],
  "partialFailedTagEvents": [ 6, 10 ],
  "failedTags": [
    {
      "tagInfo": {
        "name": "GA4",
        "executionTime": 1231,
        "id": 34,
        "status": "failure"
      },
      "eventIndex": 0
    }
  ]
}
```


---

Copyright 2021 Square, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md

### 1.0.0 / May 4th, 2021
 - Initialize

code.js
/*
Copyright 2021 Square Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/* eslint-disable eslint-comments/disable-enable-pair */
/* eslint-disable prefer-template */
/* eslint-disable object-shorthand */
/* global data */


const JSON = require('JSON');
const runContainer = require('runContainer');
const claimRequest = require('claimRequest');
const returnResponse = require('returnResponse');
const logToConsole = require('logToConsole');
const getRequestBody = require('getRequestBody');
const getRequestHeader = require('getRequestHeader');
const getContainerVersion = require('getContainerVersion');
const addEventCallback = require('addEventCallback');
const setResponseBody = require('setResponseBody');
const setResponseStatus = require('setResponseStatus');

const containerVersion = getContainerVersion();
const previewMode = containerVersion.previewMode;
const containerId = containerVersion.containerId;

const auth = getRequestHeader('Authorization');
const prodAuthCondition = !data.useBearerToken || auth === 'Bearer ' + data.bearerToken;
const previewAuthCondition = !data.usePreviewBearerToken || auth === 'Bearer ' + data.previewBearerToken;

let numSchemaErr = 0;
const invalidSchemaEvents = [];
const fullyFailedTagEvents = [];
const partialFailedTagEvents = [];
const failedTags = [];

/** **** Code execution starts here ******/

// IIFE to take advantage of return statement to end execution
(function runClient(){

  const validAuth = prodAuthCondition || (previewAuthCondition && previewMode);

  if (!validAuth) {

    if (data.rejectUnauthorized && typeof auth === 'string') {
      claimRequest();
      const error = formatAndLogErrorMessage('ERR_AUTH');
      respond(401, error);
    }

    return;
  }

  claimRequest();

  const bodyStr = getRequestBody();
  if (!bodyStr) {
    const error = formatAndLogErrorMessage('ERR_BODY_NO_JSON');
    respond(400, error);
    return;
  }

  const events = JSON.parse(bodyStr);
  if (!events || !isArray(events)) {
    const error = formatAndLogErrorMessage('ERR_BODY_ARR', { actualType: typeof events });
    respond(422, error);
    return;
  }

  // edge case for if events in payload are empty
  // unlikely to occur, but this is considered a success
  if (events.length === 0) {
    respond(200);
    return;
  }

  processEvents(
    events,
    function finishProcessing(
      totalTags,
      tagSuccess,
      tagFailure,
      eventSchemaSuccess,
      eventSchemaFailure
    ) {
      if (tagFailure === 0 && eventSchemaFailure === 0) {
        respond(200);
        return;
      }

      if (eventSchemaSuccess === 0) {
        const error = formatAndLogErrorMessage('ERR_EVENT_SCHEMA');
        respond(422, error);
        return;
      }

      if (tagSuccess === 0 && totalTags > 0) {
        const error = formatAndLogErrorMessage('ERR_TAG', { numTriggered: totalTags, failedTags: failedTags });
        respond(500, error);
        return;
      }

      const error = formatAndLogErrorMessage(
        'ERR_PARTIAL_CONTENT',
        {
          totalTags: totalTags,
          tagSuccess: tagSuccess,
          tagFailure: tagFailure,
          failedTags: failedTags,
          eventSchemaSuccess: eventSchemaSuccess,
          eventSchemaFailure: eventSchemaFailure,
        }
      );
      respond(206, error);
    }
  );
})();


/** ********* Helper methods  ***********/

function respond(statusCode, errorMessage) {
  const body = JSON.stringify({
    status: statusCode,
    error: errorMessage || '',
    invalidSchemaEvents: invalidSchemaEvents,
    fullyFailedTagEvents: fullyFailedTagEvents,
    partialFailedTagEvents: partialFailedTagEvents,
    failedTags: failedTags,
  });

  setResponseStatus(statusCode);
  setResponseBody(body);
  returnResponse();
}

// Async process method
function processEvents(events, finishProcessing) {
  const numEvents = events.length;
  let eventsCompleted = 0;
  let countTagSuccess = 0;
  let countTagFailure = 0;
  let countTotalTags = 0;
  let countEventSchemaSuccess = 0;
  let countEventSchemaFailure = 0;

  events.forEach((event, eventIndex) => {
    if (eventMatchesSchema(event, eventIndex)) {
      countEventSchemaSuccess++;
    } else {
      countEventSchemaFailure++;
      invalidSchemaEvents.push(eventIndex);

      eventsCompleted++;
      if (numEvents === eventsCompleted) {
        finishProcessing(
          countTotalTags,
          countTagSuccess,
          countTagFailure,
          countEventSchemaSuccess,
          countEventSchemaFailure
        );
      }

      return;
    }

    runContainer(event, /* onComplete= */ (bindToEvent) => {
      bindToEvent(addEventCallback)((containerId, eventData) => {
        let tagFailForEvent = 0;

        eventData.tags.forEach(tag => {
          countTotalTags++;
          if (tag.status === 'success') {
            countTagSuccess++;
          } else if (tag.status === 'failure') {
            failedTags.push({
              tagInfo: tag,
              eventIndex: eventIndex,
            });
            countTagFailure++;
            tagFailForEvent++;
          }
        });

        if (eventData.tags.length > 0) {
          // if tags fail, put event in the appropriate bucket
          if (tagFailForEvent === eventData.tags.length) {
            fullyFailedTagEvents.push(eventIndex);
            logToConsole('All ' + tagFailForEvent + ' tags failed for the ' + event.event_name + ' event');
          } else if (tagFailForEvent > 0) {
            partialFailedTagEvents.push(eventIndex);
            logToConsole(tagFailForEvent + ' / ' + eventData.tags.length + ' tags failed for the "' + event.event_name + '" event');
          }
        }

        eventsCompleted++;
        logPreview('"' + event.event_name + '" event successfully processed');
        logPreview(eventsCompleted + ' / ' + numEvents + ' events processed');

        if (numEvents === eventsCompleted) {
          finishProcessing(
            countTotalTags,
            countTagSuccess,
            countTagFailure,
            countEventSchemaSuccess,
            countEventSchemaFailure
          );
        }
      });
    });
  });
}


/** ******* Schema Definition *********/

function eventMatchesSchema(value, eventIndex) {
  const eventSchema = {
    type: 'object',
    required: 'true',
    value: {
      client_id: { type: 'string', required: false },
      event_name: { type: 'string', required: true },
      ip_override: { type: 'string', required: false },
      language: { type: 'string', required: false },
      page_encoding: { type: 'string', required: false },
      page_hostname: { type: 'string', required: false },
      page_location: { type: 'string', required: false },
      page_path: { type: 'string', required: false },
      page_referrer: { type: 'string', required: false },
      page_title: { type: 'string', required: false },
      screen_resolution: { type: 'string', required: false },
      user_agent: { type: 'string', required: false },
      user_data: {
        type: 'object',
        required: false,
        value: {
          email_address: { type: 'string', required: false },
          phone_number: { type: 'string', required: false },
          first_name: { type: 'string', required: false },
          last_name: { type: 'string', required: false },
          street: { type: 'string', required: false },
          city: { type: 'string', required: false },
          region: { type: 'string', required: false },
          postal_code: { type: 'string', required: false },
          country: { type: 'string', required: false },
        }
      },
      user_id: { type: 'string', required: false },
      viewport_size: { type: 'string', required: false },
    }
  };

  if (isArray(data.primitiveTypes)) {
    data.primitiveTypes.forEach(field => {
      eventSchema.value[field.field] = { type: field.type, required: field.required };
    });
  }

  if (data.structuralTypes) {
    const fields = JSON.parse(data.structuralTypes);
    if (fields) {
      for (const key in fields) {
        eventSchema.value[key] = fields[key];
      }
    }
  }

  return validate(value, eventSchema, 'events[' + eventIndex + ']');
}


/** ****** Schematizing methods  *******/

function validate(inputVal, schema, path) {
  const expectedType = schema.type;
  const required = schema.required;
  const schemaVal = schema.value;

  const actualType = typeof inputVal;

  if (isNullOrUndef(inputVal)) {
    if (required) {
      formatAndLogErrorMessage('ERR_MISSING', { value: inputVal, name: path });
      return false;
    } else {
      return true;
    }
  }

  if (isArray(inputVal) && expectedType === 'array') {
    const results = inputVal.map((ele, i) => {
      return validate(ele, schemaVal, path + '[' + i + ']');
    });
    return allValid(results);
  }

  if (actualType !== expectedType) {
    formatAndLogErrorMessage('ERR_SCHEMA', {
      value: inputVal,
      expectedType: expectedType,
      actualType: isArray(inputVal) ? 'array' : actualType,
      name: path,
      required: required
    });
    return false;
  }

  if (expectedType === 'object') {
    const results = [];
    for (const keyName in schemaVal) {
      results.push(validate(inputVal[keyName], schemaVal[keyName], path + '.' + keyName));
    }
    return allValid(results);
  }

  return true;
}

function allValid(arr) {
  return arr.reduce((acc, val) => acc && val, true);
}

function isNullOrUndef(value) {
  return value === undefined || value === null;
}

// built in Array.isArray method is not available in Sandboxed JS
function isArray(value) {
  return typeof value === 'object' && typeof value.length === 'number';
}


/** ********* Error handling and logging ***********/

function formatAndLogErrorMessage(errorCode, errorData) {
  const errorMessage = generateErrorMessage(errorCode, errorData);
  const fullErrorMessage = '[' + containerId + '][Server-to-Server Client][Error] - ' + errorMessage;
  logToConsole(fullErrorMessage);
  return fullErrorMessage;
}

function generateErrorMessage(name, errData) {
  const complexErrors = {
    ERR_TAG: formatTagError,
    ERR_SCHEMA: formatSchemaError,
    ERR_MISSING: formatMissingError,
    ERR_BODY_ARR: formatBodyError,
    ERR_PARTIAL_CONTENT: formatPartialContentError
  };

  if (complexErrors[name]) {
    return complexErrors[name](errData);
  }

  const simpleErrors = {
    ERR_AUTH: 'Authorization: The authorization is invalid. Ensure you are including the correct authorization header in your request.',
    ERR_BODY_NO_JSON: 'Body: The request body is missing. Verify your request body is valid JSON.',
    ERR_EVENT_SCHEMA: 'Events: No events in the request body fit the schema. Verify your request matches the common event data schema https://developers.google.com/tag-manager/serverside/common-event-data or the type checking schema you defined in the client configuration.'
  };

  if (!simpleErrors[name]) {
    return 'Error message is unspecified';
  }

  return simpleErrors[name];
}

function formatPartialContentError(errData) {
  return 'Partial Success: At least one event had the correct schema and at least one tag executed successfully. ' +
  'However, some failed:\n' +
  errData.eventSchemaSuccess + ' tag' + formatPlural(errData.eventSchemaSuccess) + 'matched the event schema.\n' +
  errData.eventSchemaFailure + ' tag' + formatPlural(errData.eventSchemaFailure) + 'did not match the event schema.\n' +
  errData.totalTags + ' tag' + formatPlural(errData.totalTags) + (errData.totalTags === 1 ? 'was ' : 'were ') + 'triggered in total.\n' +
  errData.tagSuccess + ' tag' + formatPlural(errData.tagSuccess) + 'succeeded in sending.\n' +
  errData.tagFailure + ' tag' + formatPlural(errData.tagFailure) + 'failed in sending.\n' +
  'Verify your request matches the common event data schema https://developers.google.com/tag-manager/serverside/common-event-data or the type checking schema you defined in the client configuration.\nFailed Tags Info: ' +
  JSON.stringify(errData.failedTags);
}

function formatTagError(errData) {
  return 'Tags: ' + errData.numTriggered + ' tag' + formatPlural(errData.numTriggered) +
  (errData.numTriggered === 1 ? 'was ' : 'were ') + 'triggered. ' +
  '0 tags succeeded. Check the logic in your tags to determine why they failed.\nFailed Tags Info: ' + JSON.stringify(errData.failedTags);
}

function formatBodyError(errData) {
  numSchemaErr++;

  return 'Body: The request body is not of type array. ' +
  'Instead received a type of "' + errData.actualType + '" ' +
  'The JSON may be invalid. The sGTM client expects an array of event objects.\n' +
  'Type error #' + numSchemaErr;
}

function formatSchemaError(errData) {
  numSchemaErr++;

  return 'Schema: The "' + errData.name + '" field was not of expected type "' + errData.expectedType + '". ' +
  'Instead received the value type "' + errData.actualType + '". ' +
  'This field is' + (errData.required ? ' ' : ' not ') + 'required.\n' +
  'The following value was received: ' + JSON.stringify(errData.value) + '\n' +
  'Type error #' + numSchemaErr;
}

function formatMissingError(errData) {
  numSchemaErr++;

  return 'Missing Field: The "' + errData.name + '" value is required but was not found. ' +
  'Instead received a value of "' + errData.value + '"\n Type error #' + numSchemaErr;
}

function formatPlural(num) {
  return num === 1 ? ' ' : 's ';
}

function logPreview(log) {
  if (previewMode) {
    logToConsole('[' + containerId + '][Server-to-Server Client][Info] - ' + log);
  }
}
